### ISSUE TITLE
Menu - Cannot change URL's in sub item per language

### ISSUE THREAD

**Issue by mediafish** [qTranslate-Team/qtranslate-x#120]
Hi, 

In Appearance > Menus I have one menu. 

Now I created sub Custom sub item with URL's but I cannot change the URL's per language - when changing one URL it applies to all languages which is not what I would expect. 

I created various ID's on my page - in this case on the page web-design for example: Hosting

In the English version the URL to it would be
en/web-design/#hosting
In the Spanish version the URL to it would be
/diseno-web/#hosting
In the German Version the URL to it would be
de/webdesign/#hosting

But it does not let me create different URL's and save them in the Menu Structure > Custom sub item. 
Now I can create different Navigation Labels and Title Attributes for each language and it saves them but it does not save the different URL's... which is very strange. 

I deactivated all plugins except qtranslate-x and activated the theme 2014 having this same result.

**Comment by johnclause** [comment]
Hi, please, read FAQ https://wordpress.org/plugins/qtranslate-x/faq/, topic "How can I prevent URL of a custom menu item from being converted".

Please, take the latest version from github, like https://github.com/qTranslate-Team/qtranslate-x/archive/3.2.9.8.3.zip, as 'setlang=no' feature was broken in some of earlier versions.

Also, for Spanish you need 'es/diseno-web/#hosting', otherwise it will not switch language: https://qtranslatexteam.wordpress.com/2015/02/26/browser-redirection-based-on-language/.

I am closing it now, to save clicks, we can continue discussion under closed issue, or can re-open it, if needed.

**Comment by mediafish** [comment]
Hi, 

I do not think you understand my  issue. As I have no clue what this has to do with the FAQ question you are linking to. 

How can I have different URL's itn the menu sub items? One per language? I cannot save those different URL's per language. 

I cannot save the URL's it takes one URL for all languages. Please try on my dev site.

**Comment by mediafish** [comment]
Hi again, 

I have the latest version installed. 

I go to: Appearance / Menus / English Tab
In the Custom sub item URL I add: en/web-design/#hosting
I hit Save Menu
I go to Spanish Tab
In the Custom sub item URL I add: es/diseno-web/#hosting
I hit Save Menu
Now I hit the English Tab again to see the URL and it changed the URL to: es/diseno-web/#hosting

Conclusion: I cannot add different URL's per langauge in the menu but I must be able to do this as those are the basics of a multilingual site or not?

So I do not know why you are closing this thread as for me it is a bug. Please explain.

Thanks

**Comment by Grafcom** [comment]
explained one and another here:
https://wordpress.org/support/topic/cannot-change-permalinks-1?replies=12

**Comment by mediafish** [comment]
It is explained that it is not working right now ;-)

**Comment by johnclause** [comment]
Ok, indeed, it seems I misread the issue.
>But it does not let me create different URL

This is right, the URL field is not multilingual, and it is not needed to be. Put basic bare URL there and conversion to active language happens automatically when menu is rendered on front-end. Is that what you need? The FAQ I cited, explains how to cancel this behaviour if needed.

**Comment by mediafish** [comment]
Hi, 

Thanks yes it creates urls... but not SEF urls in the frontend but this porblem alredy starts when creating pages. 

I create a page and call it 

English: Photography
German: Fotografie
Spanisch: Fotografía 

Now the URL for those pages should be: 

English: mysite.com/en/photography
German: mysite.com/de/fotografie
Spanish: mysite.com/es/fotografía 

But this is not the case! 

It creates the URL's

English: mysite.com/en/photography
German: mysite.com/de/photography
Spanish: mysite.com/es/photography

So as you see it creates SEF pages but NOT translated photography stays always instead of translating it.

And this is due to: When creating pages I cannot change the permalinks of those pages - or I can but only for one language. So when I change the permalink on the Spanish Tab of "Photography" which in Spanish is Fotografía and save it it saves it for all Languages instead of only for one language each time. 

Now to translate those URLs to their correspondent language "qTranslate slug" Plugin was made and if I use it then yes I can add the slugs to each language and my URL's look as they should: 

English: mysite.com/en/photography
German: mysite.com/de/fotografie
Spanish: mysite.com/es/fotografía 

HOWEVER:

When I do the translations with  "qTranslate slug"  it does not create the slugs if you add a Custom sub item to "Photography"

It then again creates the non translated URL's such as

English: mysite.com/en/photography/#sub-item
German: mysite.com/de/photography/#sub-item
Spanish: mysite.com/es/photography/#sub-item

instead of expected:

English: mysite.com/en/photography/#sub-item-english
German: mysite.com/de/fotografie/#sub-item-deutsch
Spanish: mysite.com/es/fotografía/#sub-item-espanol

It takes you to the correct place but with a URL that is NOT Translated. 

I believe that if you made the native wordpress permalinks inside the Pages translatable it would already solve the whole problem - but that is just a guess... and if you could solve the problem from that point the third party plugin "qTranslate slug" would not be needed and nobody would depend on their dev's...

**Comment by Grafcom** [comment]
Again, qTranslate Slug causes problems.

A workaround with SEF urls is possible. you say "But this is not the case!  It creates the URL's"
Yes, therefore, you must create them manually.  See my example with hosting. Create manual links in your menu!
Full URLs:
English: http://www.mysite.com/en/photography
German: http://www.mysite.com/de/fotografie
Spanish: http://www.mysite.com/es/fotografía
And these are still SEF urls.
And let the Navigation Labels empty on the languages you do not want to see the link.

**Comment by johnclause** [comment]
@mediafish: this should be addressed to QTranslate Slug plugin, not here. We used to be in touch with them, but they went silent: https://github.com/not-only-code/qtranslate-slug/issues/108. Unless they start acting again, we plan to re-implement -slug functionality under q-X natively, and then this problem will be gone. Until then, I suggest not to use QTranslate Slug, since the problem you address here is not the only one, you will discover more as you go deeper.

BTW, if we do this, it will be done in this way:

English: http://www.mysite.com/photography
German: http://www.mysite.com/fotografie
Spanish: http://www.mysite.com/fotografía

since slug defines the language, unless it is spelled the same, then /xx/ language code will be added.

**Comment by johnclause** [comment]
BTW, @mediafish and @Grafcom, what would you think is better, to add /xx/ for slugs spelled the same, or to require from user to always spell them differently?

**Comment by mediafish** [comment]
Hi, 

Thanks but I do not want to talk to qTranslate Slug... as I think qTranslate-X should offer SEF URL's in their own Plugin - as the Plugin is Good and should really offer this - I think it is basic to have URL's translated to their according language.... 

The URL's must be: 

English (If main language of the site no /en/ is shown at least on my site and this is good...): http://www.mysite.com/photography
German: http://www.mysite.com/de/fotografie
Spanish: http://www.mysite.com/es/fotografía

AGAIN: I think if you fixed that we can translate Permalinks inside pages  - the problem would be gone...

The user should be able to add those in the according language

@johnclause what do you mean by: "slugs spelled the same" ?

**Comment by Grafcom** [comment]
@johnclause
- what would you think is better, to add /xx/ for slugs spelled the same, or to require from user to always spell them differently? - 

/xx/ should just be the language identifier, as is now also, I think

**Comment by mediafish** [comment]
@johnclause 

yes language identifier should stay as now... if that is the question

**Comment by johnclause** [comment]
>AGAIN: I think if you fixed that we can translate Permalinks inside pages for each language - the problem would be solved...

I do not get what you are saying here ... Slugs currently come from QTranslate Slug plugin, and that plugin has to be modified to make it work correctly with q-X. If we re-implement it, we will have to also provide an import function to import the slug data from Q-Slug to q-X.

>what do you mean by: "slugs spelled the same" ?

In most of the cases slug will be different for each language, and then language code /xx/ becomes redundant in the url, as many people noticed. Then it makes sense to drop language code like '/xx/' from url, since slug already defines the language - it will look nicer then. I think there is no disagreement on that, everybody to whom I mentioned it, agreed, do you think so?

However, when in rare case two languages would have the same spelling (for example, british and american can have that easily), then we have to decide, whether we force users to make slugs different anyway or to put back /xx/ language code in url.

I personally is not strongly decided, but I am inclined to the forcing users to use different urls, since this gives them opportunity to do something more creative than robotic /xx/. They can still do 'slug-en' or something like this, which resembles the robotic way too, if nothing else is better.

What do you think?

**Comment by Grafcom** [comment]
@johnclause
- In most of the cases slug will be different for each language, and then language code /xx/ becomes redundant in the url, as many people noticed.
- I think there is no disagreement on that, everybody to whom I mentioned it, agreed, do you think so?

No, I disagree.

For example: mediafish uses mysite.com/de/fotografie
in Dutch it would it be mysite.com/nl/fotografie

photography (fotografie) is the same in both languages.
In the Netherlands we use Home for the Home page also as in English. German is Startseite (I love that language)

More and more English words are as a name and then this goes wrong.

Art Journals for example is an understanding (concept)

That's what Marit :-) and I think.

**Comment by johnclause** [comment]
I do not see how what you are saying contradicts to the fact that URL like

>http://www.mysite.com/fotografía

looks better than
>http://www.mysite.com/es/fotografía

while they both define the language? What is it exactly you disagree with? I think there is a misunderstanding here. Besides, I remember you, @Grafcom, about a month ago, asking why do we need extra /es/ in a case like this ;)

In any case, I guess we will have options for users to define how they want it.

**Comment by mediafish** [comment]
Good morning, 

I have worked on multilingual sites for over 10 years now - first time on WP and wow this really is an experience... 

Here my experience and known standards for multilingual sites:

1.) We must keep the language codes - there is no doubt about that. 
a) Now for the main language it does not show the language code - which is fine (However Joomla for example offers the option to chose if you want to show the main language with language code or without)
This is due to the fact that we have country specific domains for example mine ends in mysite.es and my language structure looks like this: 

Spanish: mysite.es/fotografia
English: mysite.es/en/photography
German: mysite.es/de/fotografie

So as you see there is no need to add a language code to the Spanish version due to its domain ending. 

Now if you had a site such as: mysite.com and you wanted your main language in Spanish - it would be nice to have the option to add the language code to your site - however this is optional but the URL would be more SEF

Spanish: mysite.com/es/fotografia
English: mysite.com/en/photography
German: mysite.com/de/fotografie

2.) @johnclause AGAIN: I think if you fixed that we can translate Permalinks inside pages for each language - the problem would be solved...

I am talking about the WP integrated Permalinks inside Pages - I am not talking about qTranslateSlug

You should make those Permalinks Translatable and then nobody would rely on a third party Plugin like qTranslateSlug

Goto: Pages / Yourpage

Language Tabs on Top

Yourpage Title
   Permalink: http://mysite.com/yourpage Edit

When you click the edit button you can change the link - this is WP integrated!

Now if I could change it and then save it per language this would be the solution. But it needs to be integrated. 

3. The same for Post / Categories / Add new Category

We can translate the name of the Category but not the Slug of each language - this must be translatable as well on a well working multilingual Plugin...

So those are the main problems I see - there are certain standards - and please do not take the language codes out - EVER. Leave it as an option for the users if they want to show language code in their main language or not - but do not take them out as this would be re-inventing the whole wheel and we do not need that...

Thanks

**Comment by Grafcom** [comment]
@mediafish,
- (However Joomla for example offers the option to chose if you want to show the main language with language code or without)

qTranslate also, see Settings - Languages - Advanced Settings - URL Modification Mode - Hide URL language information for default language. You can choose.

**Comment by johnclause** [comment]
@mediafish  thanks for the input - very valuable :+1: 

Item 1) was already addressed by @Grafcom, indeed we have that option too.

>2) You should make those Permalinks Translatable
Ok, now I understand what you meant. This is exactly what we are going to do if we re-implement -slug under -x. There is also "Slug" item, when you pull "Screen Options" down, and that one will also respond to Language Switching Buttons. However, it requires a lot of coding due to specific of slug being a part of url, it is not similar to any other translatable field. Look how much code QTranslate Slug has, this is what we would have to add approximately. Having Q-Slug plugin as an example makes it much easier, but still, it is a lot of work. I am not even sure, we may implement it as a separate plugin too.

**Comment by mediafish** [comment]
@Grafcom - thanks ok - I was not sure if it did or if it was by default only know the Plugin for 2 weeks.

@johnclause - good :-) I understand this is a free plugin and  requires a lot of coding... I just hoped your Plugin could do this already... as I am really fed up with WPML which is far too blown up I find... let's hope you will implement it one day - or WP adds multilingual use to it's core...

**Comment by LC43** [comment]
hi, sorry for being silent :)

yes, qts is big.. a big mess! :D but its working, and i've just updated qts language widget to play nice with qtx!

**Comment by mediafish** [comment]
@LC43 

Just updated the version to qTranslateX Version 3.2.9.9.3 and qTranslate Slug Version 1.1.16 and I can still not translate custom links in the menu - so - no need to talk ;-)

**Comment by mediafish** [comment]
Got rid of warning message as I had to access the site  with FTP the folder name must be qtranslate-x but it was qtranslate-x-master right so I changed it but only got rid of warning message - and I can still not translate sub item URL's to according language - it translates always all URL's to the same..

**Comment by koen84** [comment]
I hope to have language dependant slugs soon, this is quite important for SEO.  Now it looks like it's possible, but it's not, because it -as stated by others- simply changes things for all languages.

That said, i also agree that it should keep the language indentifiers in the URLs, because it's much more clear to the user (especially if not all content is translated) and better for SEO (definitely not force unique slugs across languages).

/de/fotografie
/nl/fotografie

should both be able to exist.

And then things like

/en/photography

**Comment by LC43** [comment]
@koen84 , we're talking about custom links in the menu.

You can do

**Comment by koen84** [comment]
@LC43 : this convo seems to be both about the menu links AND permalinks per language, no ?

**Comment by LC43** [comment]
@mediafish , you're right, custom links aren't parsed by qts. let me see if i can fix it next releasee.

@koen84, if you're not using not-only-code/qtranslate-slug   yet, you should ;)

**Comment by koen84** [comment]
Besides the URL / slug for pages and posts, it would be good if we could apply it also to categories.

**Comment by LC43** [comment]
you can

**Comment by koen84** [comment]
@LC43 : how should i make sure it uses this not-only-code/qtranslate-slug ?  Also, i read to NOT use qtranslate slug ?

**Comment by LC43** [comment]
@koen84 , qtx  was running too fast away from the original qtranslate, so things broke a lot. We were always compatible with qtranslate and mqtranslate. I've updated qts to be compatible again with qtx now that mqtranslate stop being actively developed. 

you can use it from the main wordpress site: [wordpress.org/plugins/qtranslate-slug](https://wordpress.org/plugins/qtranslate-slug/)

**Comment by mediafish** [comment]
@LC43 and @ koen84 - I started this topic and its about translating sub item URL's. 
I tested it today various times and it is not possible... so your message is not correct as you say yourself: "you're right, custom links aren't parsed by qts. let me see if i can fix it next releasee."

So thanks for posting but it does not concern this topic.

**Comment by Grafcom** [comment]
@all,
There is a way. Visit the site of mediafish = http://photoseller.mediafish.es/

In the menu, I created a link  called Hosting EN, Hosting DE  and Hosting ES depending on the selected language
language set to Spanish
Hosting ES brings you to photoseller.mediafish.es/diseno-web/#hosting

Change your language to English
Hosting EN brings you to photoseller.mediafish.es/en/web-design/#hosting

Change your language to German
and Hosting DE brings you to photoseller.mediafish.es/de/webdesign/#hosting

This is the intention right?

**Comment by mediafish** [comment]
Hi, 

Thanks yes that would work for one or two menu sub items but not if you have 10 or 20 or 50 imagine the mess... and it's certainly not the logic behind neither of the two Plugins - but a workaround...

You create one sub-item with URL plus two empty ones per language but we really need this clean. to be translatable in Apperance/Menu - 
Click language tab English URL must be translatable to English 
Click language tab Spanish URL must be translatable to Spanish
Cick language tab German URL must be translatable to German a.s.o

**Comment by Grafcom** [comment]
If you have 10 or 20 or 50 or more links in your menu which all go to parts of the same pages you could also consider to take a course how to create a user friendly website for your visitors and not just friendly for SEO...

**Comment by LC43** [comment]
i stand corrected.

**Comment by mediafish** [comment]
@Grafcom .... sorry what has this got to do with a clean Plugin that follows its logic - sorry but the way you propose to do it it's not userfriendly at all. 

I made over 50 multilingual websites and all of them are user friendly and SEO friendly this is possible since a long time - but not with qTranslate-X and qTranslate Slug so far but that is where we want it right?

**Comment by mediafish** [comment]
@LC43 - Excuse me what did you mean by: "i stand corrected" I don't understand ;-)

**Comment by LC43** [comment]
hi again, i just meant to say i was wrong :)

**Comment by johnclause** [comment]
People say that the latest QTranslate Slug is kind of working fine under qtx 3.3. @mediafish, do you also use it now?

@LC43, would you find time to test -slug under the latest qtx from GitHub ([beta of 3.4](https://github.com/qTranslate-Team/qtranslate-x))? Did you adjust testing of qtx presence by the constant QTX_VERSION or QTRANSLATE_FILE instead of hard-coded folder name "qtranslate-x"? Many people try to test development version of qtx, which gets installed in a folder like "qtranslate-master" and then QTranslate Slug used to get broken. I hope it does not break in this way anymore, does it?

**Comment by LC43** [comment]
@johnclause , i've left a comment on another bug in our plugin. is there a way to wait for your plugin to load before ours? thanks

**Comment by johnclause** [comment]
I do not think so, WP does not support the idea of controlling the order of plugin loading, as far as I understand, although some hackers make it happen, but there is no actual reason to. They have action  "plugins_loaded", which runs after all plugins are loaded. If you have a code that needs to run in certain order, just do not run it on load of file, but put it in the action 'plugins_loaded' or some other action that runs later. Plugin qtx has a lot of `add_action` or `apply_filters`, which you can use to enforce the sequence of execution too.

**Comment by johnclause** [comment]
@LC43, btw, qtx is almost ready to release native built-in slug support. It needs a few adjustments to allow non-latin alphabets in slugs and the rest is probably ready. Ah, also import/export from QTranslate Slug needs a bit more work. I would appreciate if you could help me to finish it, you are welcome to submit pulls with changes. If you decide to help, let me know, I will check-in the latest code then.

**Comment by LC43** [comment]
thanks john, i've changed the code that checks qtx.



and i would love to. Do you have a special branch that has this feature?

**Comment by johnclause** [comment]
No, I do not have branch, maybe I should, but it is closed to public yet. If you wish to try, uncomment line `require_once(QTRANSLATE_DIR.'/slugs/qtx_slug.php');` in `qtranslate.php`. It should work ok for latin slugs only. Also import of data from your plugin will be incomplete and may generate errors. Please, do not make changes yet, I have a bunch to commit, tried now, but I currently cannot sync with GitHub, because of some pulls I accepted, which modify the same files that I also have modified. I first need to finish new Date/Time handling I am finalizing now, before I can sync the latest changes with GitHub. This time it is a kind of big change which is unusual, I normally commit little pieces.

**Comment by LC43** [comment]
great. i'm fixing a few things on qts.

looking forward for your changes

**Comment by LC43** [comment]
i got a fatal error from the lastest code from GH and the code from wp.org doesn't seem to change the slug. i'll check the code and wait for your commits.

**Comment by ridom** [comment]
Hello @all,

@mediafish Just found out this issue because I was asking exactly the same thing : How can I easily specify a translated slug in the URL for each translated version a page.
To stick with your example, this exactly what I was imagining (of course keeping the language code es, en, de, fr, pt, ru....)  :
Spanish: mysite.com/es/fotografia
English: mysite.com/en/photography
German: mysite.com/de/fotografie

Almost one year later after the last message, I repost here to know if you have been able to find some way to realize it like it and to know if there is some plugin doing it.

I was using the Polylang plugin until the day before yesterday but I deactivated it as it was not valuable to manage different languages from the original page.
ANYWAY the thing is that this plugin used to had the feature to translate slug in order to have SEF URL's as described.

Any tips to do that would be appreciated !

**Comment by floopyzicer** [comment]
The site i used q-translate-x with q-translate slug does exactly that, you
can check if it is what you looking for and if that's what you want than
backup first and use q-translate slug with latest q-translate-x site is
lachambredessecrets dot com

**Comment by ridom** [comment]
In fact I read reply and other topics  saying that there was some incompatibility between qx and qs but if you tell me it's ok, I will try.
 Thank you for the advise.

**Comment by floopyzicer** [comment]
make backup first so you can easily restore if things go wrong.

**Comment by ridom** [comment]
I will, thank you!

**Comment by mediafish** [comment]
@ridom @floopyzicer 
No I gave it up and even changed to another CMS that supports multilingual in it's core which safes me a lot of headaches... 

I just had a look at floopyzicer's website and it seems to work.... 

but honestly if you work with Multilingual sites I would no recommend WP as a CMS - but that's my personal opinion based on experience with WPML and Q-translate which got abandonned for a while and is now Q-translate-X I believe - the developers of q-translate-x  do a great job in their spare time - but who guarantees you that they will still be around tomorrow...

**Comment by floopyzicer** [comment]
I tested several multi language plugins back in the time and q-translate x
was the best, just @JohnClause is very busy so i guess he put the plugin to
rest for a while if he find the time and have a nice chatter with @LC43 to
get both q-translate-x and q-translate slug in to one, the plugin will be
best in the market, for now we can only use it as is also latest version
works fine with Yoast and All in one Seo.
There are no guarantees for any plugin, but if they leave a mark than your
friend my friend we learn some fix and share it to communities and everyone
is happy.  lol

**Comment by ridom** [comment]
@mediafish For sure they do a great job and for sure they will probably not be around there tomorrow (maybe switching to another CMS!) but this is the price of free. What is the multilingual CMS you changed for?

@floopyzicer That would be perfect !

I have installed the qt slug plugin: it looks it works for the moment, thanks

I have a SEO question what is worst when having both pages in the same language:
-  Having duplicated content with these URL : 
  <br>**mysite.com/en/why-git-hub-is-great** and **mysite.com/pt/porque-git-hub-esta-muito-bom**
- Having  duplicated content with these URL : 
  <br>**mysite.com/en/why-git-hub-is-great** and **mysite.com/pt/why-git-hub-is-great**

In other words if I don't have translation available for a page or post should I keep the original title for both language page or should I translate at least the title in order to stick with the 2 letters language particule (en, pt, ru, fr, es, ...).
